geo/src/algorithm/stitch.rs

```diff
@@ -0,0 +1,582 @@
+    ///
+    /// ```
+    /// use geo::Stitch;
+    /// use geo::{Coord, Triangle, polygon};
+    ///
+    /// let tri1 = Triangle::from([
+    ///     Coord { x: 0.0, y: 0.0 },
+    ///     Coord { x: 1.0, y: 0.0 },
+    ///     Coord { x: 0.0, y: 1.0 },
+    /// ]);
+    /// let tri2 = Triangle::from([
+    ///     Coord { x: 1.0, y: 1.0 },
+    ///     Coord { x: 1.0, y: 0.0 },
+    ///     Coord { x: 0.0, y: 1.0 },
+    /// ]);
+    ///
+    /// let result = vec![tri1, tri2].stitch_together();
+    ///
+    /// assert!(result.is_ok());
+    ///
+    /// let mp = result.unwrap();
+    ///
+    /// assert_eq!(mp.0.len(), 1);
+    ///
+    /// let poly = mp.0[0].clone();
+    /// // 4 coords + 1 duplicate for closed-ness
+    /// assert_eq!(poly.exterior().0.len(), 4 + 1);
+    ///
+    /// let expected = polygon![
+    ///     Coord { x: 0.0, y: 1.0 },
+    ///     Coord { x: 0.0, y: 0.0 },
+    ///     Coord { x: 1.0, y: 0.0 },
+    ///     Coord { x: 1.0, y: 1.0 },
+    /// ];
+    ///
+    /// assert_eq!(poly, expected);
+    /// ```
+    ///
+    /// # Additional Notes
+    ///
+    /// If you want to do something more general like a
+    /// [`Boolean Operation Union`](https://en.wikipedia.org/wiki/Boolean_operations_on_polygons)
+    /// you should use the trait `BooleanOps` or `SpadeBoolops`.
+    fn stitch_together(&self) -> PolygonStitchingResult<MultiPolygon<T>>;
+}
+
+macro_rules! impl_stitch {
+    ($type:ty => $self:ident
+     fn stitch_together(&self) -> PolygonStitchingResult<MultiPolygon<T>> {
+         $($impls:tt)*
+     }) => {
+       impl<T> Stitch<T> for &[&$type]
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+                $($impls)*
+           }
+       }
+
+       impl<T> Stitch<T> for Vec<&$type>
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+               $self.as_slice().stitch_together()
+           }
+       }
+
+       impl<T> Stitch<T> for &[$type]
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+               $self.iter().collect::<Vec<_>>().stitch_together()
+           }
+       }
+
+       impl<T> Stitch<T> for Vec<$type>
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+               $self.iter().collect::<Vec<_>>().stitch_together()
+           }
+       }
+     };
+}
+
+impl_stitch! {
+    Polygon<T> => self
+    fn stitch_together(&self) -> PolygonStitchingResult<MultiPolygon<T>> {
+        let polys = self
+            .iter()
+            .map(|&poly| fix_orientation(poly.clone()))
+            .collect::<Vec<_>>();
+        let lines = polys
+            .iter()
+            .flat_map(|part| part.lines_iter())
+            .collect::<Vec<_>>();
```

Comment:
I tried that, but that's actually coming with perf losses. 

I think this is due to the compiler being smart: It can see that it can re-use the input vector allocation as the output vector allocation and probably does exactly that. I'm not sure about that in this specific case but I know that something like that is possible.